R/sort_levels.R

```diff
@@ -0,0 +1,67 @@
+#' Sort factor levels according to a user-defined function
+#'
+#' @param .f A factor (or character vector).
+#' @param .fun A function that will sort or permute the existing factor levels.
+#'   It must accept one character argument and return a character argument of
+#'   the same length as it's input.
+#'
+#' @param ... Additional arguments to `.fun`.
+#' @export
+#' @examples
+#' medieval_experiment <- factor(
+#'   c("Control", "Comfrey", "Cropleek", "Bishopwort")
+#' )
+#' medieval_experiment
+#'
+#' # Default (alphabetical) level-sorting:
+#' fct_sort_levels(medieval_experiment, sort)
+#'
+#' # Reversed ordering (equivalent to `fct_rev`):
+#' fct_sort_levels(medieval_experiment, sort, decreasing = TRUE)
+#'
+#' # Level-sorting with "Control" as the first level
+#' # (equivalent: `fct_relevel(medieval_experiment, "Control")`)
+#' fct_sort_levels(
+#'   medieval_experiment,
+#'   function(x) c("Control", sort(setdiff(x, "Control"))))
+#'
+#' # Randomised sorting of the levels:
+#' fct_sort_levels(medieval_experiment, sample)
+#'
+fct_sort_levels <- function(.f, .fun, ...) {
+  f <- check_factor(.f)
+  .fun <- rlang::as_function(.fun)
+
+  old_levels <- levels(f)
+  new_levels <- .fun(old_levels, ...)
+  if (!is.vector(new_levels)) {
+    stop(
+      paste("The result of `.fun(levels(.f))` must be a vector"),
+      call. = FALSE
+    )
+  }
+  if (length(old_levels) != length(new_levels)) {
```

Comment:
This feels a little overwrought to me. Doesn't `fct_relevel()` already give error messages?

Reply:
I don't think it provdes an error in this setting - if passed a (vector) proper subset of a factor's levels, `fct_relevel` just pulls the subset to the start of the factor's levels:
```
fct_relevel(factor(letters[1:3]), c("b", "a"))
[1] a b c
Levels: b a c
```

(I thought the error messages might be overkill, but was trying to follow the examples in your style guide)